test/unit/org/apache/cassandra/index/sai/cql/PlanWithIndexHintsTest.java

```diff
@@ -0,0 +1,902 @@
+        try
+        {
+            // disable query optimization so we can test the intersection clause limit in a more predictable way
+            QueryController.QUERY_OPT_LEVEL = 0;
+
+            beforeAndAfterFlush(() -> {
+                String query = "SELECT * FROM %s WHERE v1=0 AND v2=0 AND v3 = 0";
+
+                CassandraRelevantProperties.SAI_INTERSECTION_CLAUSE_LIMIT.setInt(3);
+                assertThatPlanFor(query, row1).usesAnyOf(idx1, idx2, idx3);
+                assertThatPlanFor(query + " WITH included_indexes={idx1}", row1).usesAtLeast(idx1).uses(3);
+                assertThatPlanFor(query + " WITH included_indexes={idx1, idx2}", row1).usesAtLeast(idx1, idx2).uses(3);
+                assertThatPlanFor(query + " WITH included_indexes={idx1, idx2, idx3}", row1).uses(idx1, idx2, idx3);
+
+                CassandraRelevantProperties.SAI_INTERSECTION_CLAUSE_LIMIT.setInt(2);
+                assertThatPlanFor(query, row1).usesAnyOf(idx1, idx2, idx3);
+                assertThatPlanFor(query + " WITH included_indexes={idx1}", row1).usesAtLeast(idx1).uses(2);
+                assertThatPlanFor(query + " WITH included_indexes={idx1, idx2}", row1).uses(idx1, idx2);
+                assertHintsExceedIntersectionClauseLimit(query + " WITH included_indexes={idx1, idx2, idx3}");
+
+                CassandraRelevantProperties.SAI_INTERSECTION_CLAUSE_LIMIT.setInt(1);
+                assertThatPlanFor(query, row1).usesAnyOf(idx1, idx2, idx3);
+                assertThatPlanFor(query + " WITH included_indexes={idx1}", row1).uses(idx1);
+                assertHintsExceedIntersectionClauseLimit(query + " WITH included_indexes={idx1, idx2}");
+                assertHintsExceedIntersectionClauseLimit(query + " WITH included_indexes={idx1, idx2, idx3}");
+
+                // test with an OR clause, so the intersection is nested
+                query = "SELECT * FROM %s WHERE (v1=0 AND v2=0 AND v3 = 0) OR v4 = 0";
+
+                CassandraRelevantProperties.SAI_INTERSECTION_CLAUSE_LIMIT.setInt(3);
+                assertThatPlanFor(query, row1).uses(idx1, idx2, idx3, idx4);
+                assertThatPlanFor(query + " WITH included_indexes={idx1}", row1).uses(idx1, idx2, idx3, idx4);
+                assertThatPlanFor(query + " WITH included_indexes={idx1, idx2}", row1).uses(idx1, idx2, idx3, idx4);
+                assertThatPlanFor(query + " WITH included_indexes={idx1, idx2, idx3}", row1).uses(idx1, idx2, idx3, idx4);
+
+                CassandraRelevantProperties.SAI_INTERSECTION_CLAUSE_LIMIT.setInt(2);
+                assertThatPlanFor(query, row1).usesAtLeast(idx4).uses(3);
+                assertThatPlanFor(query + " WITH included_indexes={idx1}", row1).usesAtLeast(idx1, idx4).uses(3);
+                assertThatPlanFor(query + " WITH included_indexes={idx1, idx2}", row1).uses(idx1, idx2, idx4);
+                assertHintsExceedIntersectionClauseLimit(query + " WITH included_indexes={idx1, idx2, idx3}");
```

Comment:
@pkolaczk was suggesting we should allow exceeding the intersection clause if indexes are explicitly included. What happened to that idea? 
https://github.com/datastax/cassandra/pull/1670#pullrequestreview-2815173198
I guess this would be for advanced users.
Maybe in the future we can add it behind a guardrail?

Reply:
Yes, I forgot about that discussion. It probably makes sense to exclude hinted indexes from that limit. I have just changed the code accordingly.